### PR TITLE
Allow canary releases

### PR DESCRIPTION
**Pull request by dsame**
**Description:**
https://github.com/actions/runner-images-internal/issues/4457

**Related issue:**
Add link to the related issue.

**Check list:**
- [ ] Mark if documentation changes are required.
- [ ] Mark if tests were added or updated to cover the changes.

### CODE REVIEW

File: src/installer.ts [hunk 1]
```diff
@@ -338,10 +370,112 @@ async function resolveVersionFromManifest(
...
+  if (version) {
+    core.debug(`matched: ${version}`);
+  } else {
+    core.debug('match not found');
+  }
+
+  return version;
+}
+
 // TODO - should we just export this from @actions/tool-cache? Lifted directly from there
```
Comment (brcrista) [inline src/installer.ts:467]: This comment isn't relevant anymore since we're changing the logic

File: src/installer.ts [hunk 2]
```diff
@@ -338,10 +370,112 @@ async function resolveVersionFromManifest(
...
+    }
+  }
+
+  if (range) {
+    versions = versions.sort((a, b) => {
+      if (semver.gt(a, b)) {
+        return 1;
+      }
+      return -1;
+    });
```
Comment (brcrista) [inline src/installer.ts:396]: There's a `compare` function you can pass to `Array.prototype.sort`: https://github.com/npm/node-semver#comparison

```suggestion
    versions = versions.sort(semver.compare));
```

File: src/installer.ts [hunk 3]
```diff
@@ -338,10 +370,112 @@ async function resolveVersionFromManifest(
...
+
+  return version;
+}
+
+function evaluateCanaryVersions(
+  versions: string[],
+  versionSpec: string
+): string {
+  let version = '';
+  let range: string | null | undefined;
```
Comment (brcrista) [inline src/installer.ts:424]:
```suggestion
  let range: string | undefined;
```

doesn't look like it's ever `null`?

**Review by brcrista — APPROVED** [review]
A few minor suggestions but looks good

**Review by marko-zivic-93 — APPROVED** [review]

File: src/installer.ts [hunk 4]
```diff
@@ -338,10 +370,112 @@ async function resolveVersionFromManifest(
   }
 }
 
+function evaluateNightlyVersions(
+  versions: string[],
+  versionSpec: string
+): string {
+  let version = '';
+  let range: string | null | undefined;
```
Comment (brcrista) [inline src/installer.ts:378]:
```suggestion
  let range: string | undefined;
```

doesn't look like it's ever `null`?

**Comment by dsame** [comment]
The PR to be abandoned for sake of this PR https://github.com/actions/setup-node/pull/619

**Comment by dmitry-shibanov** [comment]
I'm going to close the pull request in favour of this: https://github.com/actions/setup-node/pull/655

**Pull request closed without merging**